semantic/mesh/roof/mesh-repair.cpp

```diff
@@ -0,0 +1,284 @@
+        return false;
+    }
+
+    // Verify that there are not any common neighbours between the two faces
+    auto neigh1 { ffTable[fI1] };
+    std::sort(neigh1.begin(), neigh1.end());
+    auto neigh2 { ffTable[fI2] };
+    std::sort(neigh2.begin(), neigh2.end());
+    std::vector<std::size_t> commonNeighbours;
+    std::set_intersection(neigh1.begin(),
+                          neigh1.end(),
+                          neigh2.begin(),
+                          neigh2.end(),
+                          std::back_inserter(commonNeighbours));
+    if (commonNeighbours.size() > 0) { return false; }
+
+    // Construct the new faces
+    auto& f1 { mesh.faces[fI1] };
+    auto& f2 { mesh.faces[fI2] };
+    auto a { theOtherVertex(f1, v1, v2) };
+    auto b { theOtherVertex(f2, v2, v1) };
+    auto nf1 { Face(a, b, v2, f2.imageId) };
+    auto nf2 { Face(b, a, v1, f2.imageId) };
+
+    // Check they are not null
+    if (isNullFace(mesh, nf1) || isNullFace(mesh, nf2)) { return false; }
+
+    mesh.faces[fI1] = nf1;
+    mesh.faces[fI2] = nf2;
+    return true;
+}
+
+/// Flip the longest edge - works in our case
+bool removeNullFaceByEdgeFlip(Mesh& mesh, const FFTable& ffTable, const FIdx fI)
+{
+    auto& f { mesh.faces[fI] };
+
+    // flip the longest edge
+    auto lab { math::length(mesh.vertices[f.a] - mesh.vertices[f.b]) };
+    auto lbc { math::length(mesh.vertices[f.b] - mesh.vertices[f.c]) };
+    auto lca { math::length(mesh.vertices[f.c] - mesh.vertices[f.a]) };
+
+    if (lab > lbc && lab > lca)
+    {
+        return edgeFlip(mesh, ffTable, fI, f.a, f.b);
+    }
+    else
+    {
+        if (lbc > lca) { return edgeFlip(mesh, ffTable, fI, f.b, f.c); }
+        else { return edgeFlip(mesh, ffTable, fI, f.c, f.a); }
+    }
+}
+
+
+bool attemptToRemoveNullFacesByEdgeFlip(Mesh& mesh)
+{
+    std::size_t flipped { 0 };
+    auto ffTable { geometry::getFaceFaceTableNonManifold(mesh) };
+    for (FIdx fI = 0; fI < mesh.faces.size(); fI++)
+    {
+        if (isNullFace(mesh, mesh.faces[fI]))
+        {
+            if (removeNullFaceByEdgeFlip(mesh, ffTable, fI))
+            {
+                // meshes are small, OK to just recompute it all
+                ffTable = geometry::getFaceFaceTableNonManifold(mesh);
+                ++flipped;
+            }
+        }
+    }
+    LOG(info1) << "Flipped " << flipped
+               << " edges while attemping to remove null faces.";
+    return flipped > 0;
+}
+
+/// Remove faces that are adjacent to each other on all three sides
+void removeSplicedFaces(Mesh& mesh)
+{
+    auto ffTable { geometry::getFaceFaceTableNonManifold(mesh) };
+    std::vector<bool> removeFace(mesh.faces.size(), false);
+    for (FIdx fI = 0; fI < mesh.faces.size(); fI++)
+    {
+        std::map<FIdx, std::size_t> neighCnt;
+        for (auto n : ffTable[fI])
+        {
+            if (neighCnt.count(n)) { neighCnt[n]++; }
+            else { neighCnt[n] = 1; }
+        }
+
+        for (auto nc : neighCnt)
```

Comment:
chatch iterated value by `const auto&` pls (unless it's an ordinal or you need a copy, ofc)

Reply:
Sorry, fixed